packages/lexicon/src/Cookies/index.ts

```diff
@@ -0,0 +1,307 @@
+import JsCookie from 'js-cookie';
+
+const cookieTray = {
+  accessToken: 'app.at',
+  accessTokenExpiration: 'app.at_exp',
+  user: 'user',
+  lastState: 'lastState',
+  identityToken: 'app.idt',
+  refreshToken: 'app.rt',
+} as const;
+
+type Cookies = typeof cookieTray;
+
+type CookieName = Cookies[keyof Cookies];
+
+type CookieConstructor<CookieClass> = new (name: string) => CookieClass;
+type CookieKind =
+  | CookieConstructor<UserCookie>
+  | CookieConstructor<RefreshTokenCookie>
+  | CookieConstructor<LastStateCookie>
+  | CookieConstructor<AccessTokenCookie>
+  | CookieConstructor<AccessTokenExpirationCookie>
+  | CookieConstructor<IdentityTokenCookie>;
+
+export class EmptyCookie<T = CookieKind> {
+  readonly kind: CookieKind;
+
+  constructor(kind: CookieKind) {
+    this.kind = kind;
+  }
+
+  isEmpty(): this is EmptyCookie<T> {
+    return true;
+  }
+
+  isNotEmpty(): this is T {
+    return false;
+  }
+
+  fillWith(value: string): T | EmptyCookie<T> {
+    const result = JsCookie.set(this.kind.name, value);
+
+    return result ? (new this.kind(value) as T) : this;
+  }
+}
+
+export class CookieJar {
+  static set(
+    name: 'app.at',
+    value: string,
+  ): AccessTokenCookie | EmptyCookie<AccessTokenCookie>;
+  static set(
+    name: 'app.at_exp',
+    value: string,
+  ): AccessTokenExpirationCookie | EmptyCookie<AccessTokenExpirationCookie>;
+  static set(
+    name: 'app.idt',
+    value: string,
+  ): IdentityTokenCookie | EmptyCookie<IdentityTokenCookie>;
+  static set(
+    name: 'lastState',
+    value: string,
+  ): LastStateCookie | EmptyCookie<LastStateCookie>;
+  static set(
+    name: 'app.rt',
+    value: string,
+  ): RefreshTokenCookie | EmptyCookie<RefreshTokenCookie>;
+  static set(name: 'user', value: string): UserCookie | EmptyCookie<UserCookie>;
+  static set(name: CookieName, value: string) {
+    const didSetCookieSuccessfully = Boolean(JsCookie.set(name, value));
+
+    switch (name) {
+      case 'user':
+        return didSetCookieSuccessfully
+          ? new UserCookie(value)
+          : new EmptyCookie<UserCookie>(UserCookie);
+        break;
```

Comment:
It looks like these breaks are unreachable. They have red squigglies in my editor. Not hurting anything though.